Add a otelcol.auth.oauth2 component (#3143)
grafana/agent: docs/sources/flow/reference/components/otelcol.auth.oauth2.md

---
title: otelcol.auth.oauth2
---

# otelcol.auth.oauth2

> `otelcol.auth.oauth2` exposes a `handler` that can be used by other `otelcol`
> components to authenticate requests using oauth2 authentication.

Reviewer: ```suggestion
`otelcol.auth.oauth2` exposes a `handler` that can be used by other `otelcol`
components to authenticate requests using OAuth 2.0.
```

The authorization tokens can be used by HTTP and gRPC based OpenTelemetry exporters. 
This component can fetch and refresh expired tokens automatically. For further details about 

> OAuth2 Client Credentials flow (2-legged workflow) see [here](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4).

Reviewer: The official name seems to be "OAuth 2.0": https://oauth.net/2/

```suggestion
OAuth 2.0 Client Credentials flow (2-legged workflow) see [here](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4).
```

> **NOTE**: `otelcol.auth.oauth2` is a wrapper over the upstream OpenTelemetry
> Collector `oauth2client` extension. Bug reports or feature requests will be
> redirected to the upstream repository, if necessary.

Multiple `otelcol.auth.oauth2` components can be specified by giving them
different labels.

## Usage

```river
otelcol.auth.oauth2 "LABEL" {
    client_id     = "CLIENT_ID"
    client_secret = "CLIENT_SECRET"
    token_url     = "TOKEN_URL"
}
```

## Arguments

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`client_id` | `string` | The client identifier issued to the client. | | yes
`client_secret` | `string` | The secret string associated with the client identifier. | | yes
`token_url` | `string` | The server endpoint URL from which to get tokens. | | yes

> `endpoint_params` | `map([]string)` | Additional parameters that are sent to the token endpoint. | `{}` | no

Reviewer: ```suggestion
`endpoint_params` | `map(list(string))` | Additional parameters that are sent to the token endpoint. | `{}` | no
```

> `scopes` | `[]string` | Requested permissions associated for the client. | `[]` | no

Reviewer: ```suggestion
`scopes` | `list(string)` | Requested permissions associated for the client. | `[]` | no
```

> `timeout` | `duration` | The timeout on the client connecting to `token_url` for both initial tokens and refreshing tokens. `"0s"` implies no timeout. | `"0s"` | no

Reviewer: Can you make this shorter and move the longer description outside of the table? The description in the table should be a short description, and attribute behavior should be documented outside of it. (This is something I'll document)

## Blocks

The following blocks are supported inside the definition of
`otelcol.auth.oauth2`:

Hierarchy | Block | Description | Required
--------- | ----- | ----------- | --------
tls | [tls][] | TLS settings for the token client. | no

[tls]: #tls-block

### tls block

The `tls` block configures TLS settings used for connecting to the token client. If the `tls` block isn't provided, 
TLS won't be used for communication.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`ca_file` | `string` | Path to the CA file. | | no
`cert_file` | `string` | Path to the TLS certificate. | | no
`key_file` | `string` | Path to the TLS certificate key. | | no
`min_version` | `string` | Minimum acceptable TLS version for connections. | `"TLS 1.2"` | no
`max_version` | `string` | Maximum acceptable TLS version for connections. | `"TLS 1.3"` | no
`reload_interval` | `duration` | Frequency to reload the certificates. | | no

> `client_ca_file` | `string` | Path to the CA file used to authenticate client certificates. | | no

Reviewer: Do we have a shared block for this? I feel like we do.

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`handler` | `capsule(otelcol.Handler)` | A value that other components can use to authenticate requests.

## Component health

`otelcol.auth.oauth2` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`otelcol.auth.oauth2` does not expose any component-specific debug information.

## Example

> This example configures [otelcol.exporter.otlp][] to use oauth2 authentication:

Reviewer: ```suggestion
This example configures [otelcol.exporter.otlp][] to use OAuth 2.0 for authentication:
```
